object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java

```diff
@@ -279,6 +306,276 @@ private static String getExactStringValue(Tree tree, ValueAnnotatedTypeFactory f
+          Node result = ((ReturnNode) node).getResult();
+          if (result instanceof LocalVariableNode
+              && isVarInPrevDefs(newDefs, (LocalVariableNode) result)) {
+            newDefs.remove(getAssignmentNodeOfVar(newDefs, (LocalVariableNode) result));
+          }
+        }
+      }
+
+      for (BlockImpl succ : getSuccessors(state.first)) {
+        Set<AssignmentNode> toRemove = new HashSet<>();
+
+        // There is nothing needs to be checked
+        if (newDefs.size() == 0) {
+          propagate(Pair.of(succ, newDefs), visited, worklist);
+          continue;
+        }
+
+        CFStore succRegularStore = getAnalysis().getInput(succ).getRegularStore();
+
+        for (AssignmentNode assignmentNode : newDefs) {
+
+          // If the successor block is the exit block or if the set of the pair of block id and set
+          // of defs is repeated in a loop or if the variable is going out of scope
+          if (succ instanceof SpecialBlockImpl
+              || visited.contains(Pair.of(succ.getId(), newDefs))
+              || succRegularStore.getValue((LocalVariableNode) (assignmentNode.getTarget()))
+                  == null) {
+
+            if (nodes.size() == 0) { // If the cur block is special or conditional block
+              reportAlwaysCallErrors(assignmentNode, succRegularStore, null);
+
+            } else { // If the cur block is Exception/Regular block then it checks AlwaysCall
+              // condition in the store right after the last node
+              Node last = nodes.get(nodes.size() - 1);
+              CFStore storeAfter = getStoreAfter(last);
+              reportAlwaysCallErrors(
+                  assignmentNode,
+                  storeAfter,
+                  (last instanceof AssignmentNode) ? getAnnotatedType(last.getTree()) : null);
+            }
+          }
+
+          if (succRegularStore.getValue((LocalVariableNode) (assignmentNode.getTarget())) == null) {
+            // If the lhs of the assignmentNode isn't in the successor's input store, it means the
+            // variable is going out of scope
+            toRemove.add(assignmentNode);
+          }
+        }
+
+        newDefs.removeAll(toRemove);
+        propagate(Pair.of(succ, newDefs), visited, worklist);
+      }
+    }
+  }
+
+  /**
+   * it inputs a set of AssignmentNodes(defs) and a LocalVariableNode(localVariableNode), if there
+   * is an AssignmentNode in defs that its lhs is equal to localVarNode, then it returns that
+   * AssignmentNode, otherwise it returns null.
+   *
+   * @param defs
+   * @param localVarNode
+   * @return
+   */
+  private AssignmentNode getAssignmentNodeOfVar(
+      Set<AssignmentNode> defs, LocalVariableNode localVarNode) {
+
+    for (AssignmentNode assignmentNode : defs) {
+      LocalVariableNode lhs = (LocalVariableNode) assignmentNode.getTarget();
+
+      if (lhs.getElement() == localVarNode.getElement()) {
+        return assignmentNode;
+      }
+    }
+
+    return null;
```

Comment:
Here I think you should throw an exception rather than just returning `null`.  We only expect this method to be called when a relevant assignment is in the `defs` set